### ISSUE TITLE
MacOS build instructions no longer mention OpenSSL

### ISSUE THREAD

**Issue by CoffeeFlux** [dotnet/runtime#1461]
To build libraries on MacOS, you have to either set an environmental variable or symlink openssl to a known location. This was documented in the old corefx repo: https://github.com/dotnet/corefx/blob/master/Documentation/building/unix-instructions.md#macos

Unfortunately, that documentation didn't make it over, so it's not obvious how to build libraries on MacOS in the new repo following instructions in the README. Ideally, this workflow would be revisited as it makes for a poor developer experience, but for now updated documentation would be great.

It appears the need for pkgconfig also did not make it over to the new dependencies list.

**Comment by jeffschwMSFT** [comment]
cc @bartonjs

**Comment by akoeplinger** [comment]
Can we make building the OpenSSL library optional on Mac so that we don't need to do these symlinks? .NET Core uses SecureTransport by default since quite some time so I guess this library is just for backwards compat?

**Comment by bartonjs** [comment]
> Can we make building the OpenSSL library optional on Mac so that we don't need to do these symlinks? 

I'd really rather not.  It makes it easy for us to mess up the official build and testing. (A more interesting question is: can we drop it altogether?).

It's only used for AesCcm, AesGcm, DSAOpenSsl, ECDiffieHellmanOpenSsl, ECDsaOpenSsl, RSAOpenSsl (and SafeEvpPKeyHandle), but it's still used, and therefore needs to be part of the shared runtime.

**Comment by akoeplinger** [comment]
> It makes it easy for us to mess up the official build and testing.

Ok, what about an easier opt-out via some build.sh option? We could print the option when CMake doesn't find openssl.

**Comment by CoffeeFlux** [comment]
Just tried the exact commands in the old CoreFX repo (I have stock homebrew with no special config) and my build is still failing:

```
  Scanning dependencies of target System.Security.Cryptography.Native.OpenSsl
  [ 83%] Linking C shared library System.Security.Cryptography.Native.OpenSsl.dylib
  Undefined symbols for architecture x86_64:
    "_X509_STORE_CTX_get0_store", referenced from:
        _CryptoNative_X509StoreCtxReset in pal_x509.c.o
        _CryptoNative_X509StoreCtxResetForSignatureError in pal_x509.c.o
        _CheckOcsp in pal_x509.c.o
       (maybe you meant: _local_X509_STORE_CTX_get0_store)
    "_X509_check_host", referenced from:
        _CryptoNative_CheckX509Hostname in openssl.c.o
  Failed to build corefx native components.
       (maybe you meant: _local_X509_check_host)
  ld: symbol(s) not found for architecture x86_64
clang : error : linker command failed with exit code 1 (use -v to see invocation) [/Users/ryan/git/runtime/src/libraries/Native/build-native.proj]
  make[2]: *** [System.Security.Cryptography.Native/System.Security.Cryptography.Native.OpenSsl.dylib] Error 1
  make[1]: *** [System.Security.Cryptography.Native/CMakeFiles/System.Security.Cryptography.Native.OpenSsl.dir/all] Error 2
  make: *** [all] Error 2
/Users/ryan/git/runtime/src/libraries/Native/build-native.proj(36,5): error MSB3073: The command ""/Users/ryan/git/runtime/src/libraries/Native/build-native.sh" x64 Debug OSX outconfig netcoreapp5.0-OSX-Debug-x64 -portable" exited with code 1.

Build FAILED.
```

**Comment by bartonjs** [comment]
@CoffeeFlux Do you possibly have a mismatch where it's the OpenSSL 1.1 headers but the OpenSSL 1.0.2 binaries?

**Comment by CoffeeFlux** [comment]
I think so, yeah. I assume the pkg-config files that are installed with the instructions in the CoreFX repo point to the homebrew installation, which nowadays is OpenSSL 1.1. The binary seems to be the system-provided one (see below). For reference, I'm on Mojave.

```
  -- Found OpenSSL: /usr/lib/libcrypto.dylib (found version "1.0.2t")  
  -- Looking for EC_GF2m_simple_method
  -- Looking for EC_GF2m_simple_method - found
  -- Looking for SSL_get0_alpn_selected
  -- Looking for SSL_get0_alpn_selected - found
  -- Configuring done
```

**Comment by bartonjs** [comment]
Huh, it certainly shouldn't have picked the /usr/lib copy (which is LibreSSL).

There's a current PR that makes this simpler on macOS, largely because of the OpenSSL 1.0.2 End of Life, but it'll still make it easier :smile:.

**Comment by CoffeeFlux** [comment]
As of your latest PR, adding this to my bash_profile seems to make CoreFX build without issue:

```
# openssl is keg-only
export PATH="/usr/local/opt/openssl@1.1/bin:$PATH"
export LDFLAGS="-L/usr/local/opt/openssl@1.1/lib"
export CPPFLAGS="-I/usr/local/opt/openssl@1.1/include"
export PKG_CONFIG_PATH="/usr/local/opt/openssl@1.1/lib/pkgconfig"
```

**Comment by BruceForstall** [comment]
@bartonjs 

I tried this:
```
$ brew install cmake pkgconfig openssl
```

It turns out I already had all these installed. Then (I augmented the comments here to be more descriptive):
```
# We need to make the runtime libraries discoverable, as well as make
# pkg-config be able to find the headers and current ABI version.
#
# Ensure the paths we will need exist
mkdir -p /usr/local/lib/pkgconfig

# The rest of these instructions assume a default Homebrew path of
# `/usr/local/opt/<module>`, with `brew --prefix` returning `/usr/local`
# and `brew --prefix openssl` returning `/usr/local/opt/openssl@1.1`.
# In this case, `brew info openssl` shows:
# `openssl@1.1: stable 1.1.1d (bottled) [keg-only]`.
#
# Runtime dependencies
ln -s /usr/local/opt/openssl\@1.1/lib/libcrypto.1.1.dylib /usr/local/lib/
ln -s /usr/local/opt/openssl\@1.1/lib/libssl.1.1.dylib /usr/local/lib/

# Compile-time dependencies (for pkg-config)
ln -s /usr/local/opt/openssl\@1.1/lib/pkgconfig/libcrypto.pc /usr/local/lib/pkgconfig/
ln -s /usr/local/opt/openssl\@1.1/lib/pkgconfig/libssl.pc /usr/local/lib/pkgconfig/
ln -s /usr/local/opt/openssl\@1.1/lib/pkgconfig/openssl.pc /usr/local/lib/pkgconfig/
```

However, I still got the same error with root-level `./build.sh`.

**Comment by bartonjs** [comment]
Did you clean the build?  Sometimes cmake remembers things that it shouldn't have.

**Comment by BruceForstall** [comment]
@bartonjs Yes, every time I do a build, I run `git clean -dxf`, then recycle the Terminal window.

I also reverted those `ln` changes, and tried what @CoffeeFlux suggested above (adding this to .bash_profile):
```
# openssl is keg-only
export PATH="/usr/local/opt/openssl@1.1/bin:$PATH"
export LDFLAGS="-L/usr/local/opt/openssl@1.1/lib"
export CPPFLAGS="-I/usr/local/opt/openssl@1.1/include"
export PKG_CONFIG_PATH="/usr/local/opt/openssl@1.1/lib/pkgconfig"
```
and that did not work for me, either; I get the same build failure.

**Comment by BruceForstall** [comment]
@bartonjs I reverted all these changes (so, no `ln`, no export of LDFLAGS/CPPFLAGS/PKG_CONFIG_PATH), and followed the cmake error advice and added the following to .bash_profile:
```
export OPENSSL_ROOT_DIR="/usr/local/opt/openssl@1.1/"
```
and then the build succeeded.

Is this the appropriate advice to document?

**Comment by bartonjs** [comment]
No, it shouldn't require editing .bash_profile.

My best guess (I haven't been able to mess with a mac yet) is that pkgconfig isn't finding the .pc files.

There's some syntax like `pkgconifg ---include openssl` (which you might need to say `pkgconfig --include libcrypto`, because the openssl one is just a "I have nested needs" and I don't remember how the commandline tool finds it).

If you cat the /usr/local/lib/pkgconfig/*.pc files, do they all cat successfully and have "valid looking" data?

**Comment by akoeplinger** [comment]
If you want to use the `OPENSSL_ROOT_DIR` approach you don't need to put it in your .bash-profile, it's enough for the env variable to be set during the build.

E.g. doing `OPENSSL_ROOT_DIR="/usr/local/opt/openssl@1.1" ./build.sh` will work.

**Comment by CoffeeFlux** [comment]
Yeah, I think setting the env var is probably the best advice to document unless there's a compelling reason to use the symlinks over it. If someone needs it set perpetually they can add it to their bash_profile, and otherwise it's the simplest solution.

**Comment by bartonjs** [comment]
The expected state for pkgconfig:

```
$ pkg-config --cflags --libs openssl
-I/usr/local/Cellar/openssl@1.1/1.1.1d/include -L/usr/local/Cellar/openssl@1.1/1.1.1d/lib -lssl -lcrypto
```

After `git clean -xdf` I successfully built from `src/libraries/Native` with `./build-native.sh`

I have no custom environment variables set, here's the state of the relevant files:

```
$ ls -al /usr/local/lib/pkgconfig/
total 0
drwxr-xr-x   9 jbarton  admin  288 Jan  9 11:03 .
drwxrwxr-x  14 jbarton  admin  448 Jan  9 11:26 ..
lrwxr-xr-x   1 jbarton  admin   49 Jun 28  2018 libcrypto.pc -> /usr/local/opt/openssl/lib/pkgconfig/libcrypto.pc
lrwxr-xr-x   1 jbarton  admin   46 Jan  9 11:03 liblzma.pc -> ../../Cellar/xz/5.2.4/lib/pkgconfig/liblzma.pc
lrwxr-xr-x   1 jbarton  admin   46 Jun 28  2018 libssl.pc -> /usr/local/opt/openssl/lib/pkgconfig/libssl.pc
lrwxr-xr-x   1 jbarton  admin   47 Jun 28  2018 openssl.pc -> /usr/local/opt/openssl/lib/pkgconfig/openssl.pc
lrwxr-xr-x   1 jbarton  admin   55 Jan  9 10:59 python-3.7.pc -> ../../Cellar/python/3.7.6_1/lib/pkgconfig/python-3.7.pc
lrwxr-xr-x   1 jbarton  admin   56 Jan  9 10:59 python-3.7m.pc -> ../../Cellar/python/3.7.6_1/lib/pkgconfig/python-3.7m.pc
lrwxr-xr-x   1 jbarton  admin   52 Jan  9 10:59 python3.pc -> ../../Cellar/python/3.7.6_1/lib/pkgconfig/python3.pc
```

```$ cat /usr/local/lib/pkgconfig/{openssl,libssl,libcrypto}.pc
prefix=/usr/local/Cellar/openssl@1.1/1.1.1d
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include

Name: OpenSSL
Description: Secure Sockets Layer and cryptography libraries and tools
Version: 1.1.1d
Requires: libssl libcrypto
prefix=/usr/local/Cellar/openssl@1.1/1.1.1d
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include

Name: OpenSSL-libssl
Description: Secure Sockets Layer and cryptography libraries
Version: 1.1.1d
Requires.private: libcrypto
Libs: -L${libdir} -lssl
Cflags: -I${includedir}
prefix=/usr/local/Cellar/openssl@1.1/1.1.1d
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include
enginesdir=${libdir}/engines-1.1

Name: OpenSSL-libcrypto
Description: OpenSSL cryptography library
Version: 1.1.1d
Libs: -L${libdir} -lcrypto
Libs.private:  
Cflags: -I${includedir}
```

**Comment by akoeplinger** [comment]
@bartonjs I assume you manually created these .pc files (the symlinks created by homebrew point to ../../Cellar) while you have absolute paths there.

On my Mac I don't have these symlinks even though I have openssl and openssl@1.1 installed.

**Comment by bartonjs** [comment]
@akoeplinger Yeah.  Homebrew got opinionated about integrating OpenSSL, and won't do it.  In order to function at runtime you need to symlink the libraries into /usr/local/lib.  Both steps are annoying, since OpenSSL 1.0.x and OpenSSL 1.1.x are SxS with each other, the legacy OpenSSL 0.9.8 still included on the system, and the LibreSSL included on the system.

The draft steps are

```sh
# We need to make the runtime libraries discoverable, as well as make
# pkg-config be able to find the headers and current ABI version.
#
# Ensure the paths we will need exist
mkdir -p /usr/local/lib/pkgconfig

# The rest of these instructions assume a default Homebrew path of
# /usr/local/opt/<module>, with /usr/local being the answer to
# `brew --prefix`.
#
# Runtime dependencies
ln -s /usr/local/opt/openssl\@1.1/lib/libcrypto.1.1.dylib /usr/local/lib/
ln -s /usr/local/opt/openssl\@1.1/lib/libssl.1.1.dylib /usr/local/lib/

# Compile-time dependencies (for pkg-config)
ln -s /usr/local/opt/openssl/lib/pkgconfig/libcrypto.pc /usr/local/lib/pkgconfig/
ln -s /usr/local/opt/openssl/lib/pkgconfig/libssl.pc /usr/local/lib/pkgconfig/
ln -s /usr/local/opt/openssl/lib/pkgconfig/openssl.pc /usr/local/lib/pkgconfig/
```

(Note that I've intentionally used the unversioned openssl in my .pc symlinks, since that sets things up for changing to vNext whenever Homebrew wants to do that, too)

**Comment by akoeplinger** [comment]
Yeah it's kinda sad that we put all contributors through this dance even though almost no one will use OpenSSL on Mac.

**Comment by bartonjs** [comment]
Yeah, I like your opt-out idea.  I don't know that I'd plumb it through to the tests layer, but being able to tell the build to not care is reasonable.

**Comment by BruceForstall** [comment]
@bartonjs If I use your "draft steps" above, then I get:
```
~/gh/runtime$ pkg-config --cflags --libs openssl
-I/usr/local/Cellar/openssl/1.0.2t/include -L/usr/local/Cellar/openssl/1.0.2t/lib -lssl -lcrypto
```

which it seems like you wouldn't want. If I use the version -specific `ln` commands:
```
~/gh/runtime$ ln -s /usr/local/opt/openssl\@1.1/lib/pkgconfig/libcrypto.pc /usr/local/lib/pkgconfig/
~/gh/runtime$ ln -s /usr/local/opt/openssl\@1.1/lib/pkgconfig/libssl.pc /usr/local/lib/pkgconfig/
~/gh/runtime$ ln -s /usr/local/opt/openssl\@1.1/lib/pkgconfig/openssl.pc /usr/local/lib/pkgconfig/
```

then I get:
```
~/gh/runtime$ pkg-config --cflags --libs openssl
-I/usr/local/Cellar/openssl@1.1/1.1.1d/include -L/usr/local/Cellar/openssl@1.1/1.1.1d/lib -lssl -lcrypto
```
isn't this preferable?

**Comment by bartonjs** [comment]
Using the versionless ones, do you get repointed from 1.0 to 1.1 after `brew upgrade`? (That's what happened on my machine a couple weeks ago)

**Comment by bartonjs** [comment]
Also, now that we do portable binding at runtime, it doesn't matter if it finds the 1.0 or 1.1 headers :)

**Comment by BruceForstall** [comment]
Maybe I had a problem where pkg-config wasn't fully installed.

I ran:
```
brew install pkg-config
...
Warning: pkg-config 0.29.2 is already installed, it's just not linked
You can use `brew link pkg-config` to link this version.
```
I'd seen this before, when my builds were failing.

So I ran:
```
brew link pkg-config
Linking /usr/local/Cellar/pkg-config/0.29.2... 4 symlinks created
```

Now, with the script above (I used mine, above, with the pkgconfig version-specific `ln` commands), the build succeeded with a plain, root-level `./build.sh`.

> Using the versionless ones, do you get repointed from 1.0 to 1.1 after brew upgrade? (That's what happened on my machine a couple weeks ago)

I've run `brew update` and `brew upgrade` several times. Do you mean: if I create the versionless `ln` pkgconfig links, then run `brew upgrade` (again), do the links get updated to point to the version-specific (1.1) files?

**Comment by bartonjs** [comment]
When I did brew upgrade to upgrade from many many moons ago to now, it repointed `/usr/local/opt/openssl` from the 1.0 directory to the 1.1 directory.  If you're not seeing that on yours, we can certainly write the directions more explicitly.  I was aiming for "convenient", but not at the expense of correct.

**Comment by BruceForstall** [comment]
Proposed documentation change in https://github.com/dotnet/runtime/pull/1836

**Comment by bartonjs** [comment]
Accepted #1836.  Thanks, Bruce.
